readme.txt
   =====================================
    _Elite_Adventurous_  version:beta-2.0001    17/05/16
   =====================================

    < the rest of this file is no longer updated >
    <   for further changes see docs directory   >

09/04/16 London
    alpha release rc_a5
24/03/16 London
    alpha release rc_a4
17/03/16 London
    prepare alpha release (branch rc_a4)
14/03/16 London
    dx branch for directx fix (win8.1)
    
 +------------------------------------------------+
 | based on Elite - The New Kind                  |
 |  by Christian Pinder <christian@newkind.co.uk> |
 |                      http://www.newkind.co.uk  |
 +------------------------------------------------+

  Contents
  --------
    allegro - source and VS2015 build of Allegro
    assets  - sound and graphics
    config  - configuration files
    docs    - documentation
    src     - source and project files

 +---------------------------------------------------
 | The following text is the original readme.txt
 | file from the newkind repository...

Elite - The New Kind   Release 1.0
--------------------   -----------

Revision date: 22 July 2001

This is release 1.0 of Elite - The New Kind.
For changes since previous releases see below.

newkindb.zip contains a compiled version of the code designed to run under MS Windows 95/98/NT using the Allegro graphics library and DirectX.

newkind.zip contains source code for Elite - The New Kind.
If you want to recompile the game then please note the following...
a. The .wav and .dat files are not included in the source distribution to keep the size down, they can be found in newkindb.zip.
b. You need the WIP version of Allegro to compile the code.

The latest versions of the source and executable can always be found on the New Kind website...
http://www.newkind.co.uk

To run the supplied executable you will need DirectX installed.  Windows 98/ME/2000 and NT 4 (with latest service pack) come with this.
If you are using Windows 95 and you haven't already installed DirectX you will need to do so.
The DirectX runtime can be downloaded from the Microsoft website. (www.microsoft.com/directx).

Two pieces of music are included in Elite - The New Kind.  The Elite Theme and The Blue Danube.
The Elite Theme was composed by Aidan Bell and the Blue Danube was composed by Strauss.
I spent a long time hunting for the best MIDI versions of these pieces I could find.
To appreciate them properly you will need a decent Soft Synth Midi driver installed and set as you prefered Midi and Sound device.

Keys you can use...
Press Y or N on the intro screen.
Press Space on the ship parade screen.

F1  - Launch when docked, Front View when in flight.
F2  - Rear View
F3  - Left View
F4  - Right View when in flight.
    When docked, Buy Equipment for ship.
      Use up and down arrow keys to select item, return/enter key to buy.
F5  - Display Galactic Chart.
F6  - Short Range Chart.
F7  - Show information on selected planet.
F8  - Buy and sell items on the stock market.
      Use up and down arrow keys to select item, right arrow key to buy, left arrow key to sell.
F9  - Commander information.
F10 - Inventory.
F11 - Options screen (Save Game, Load Game, Game Settings, Quit).
      Use up and down arrows keys to select option, return/enter to select.

 A - Fire.
 S - Dive.
 X - Climb.
 < - Roll Left
 > - Roll Right
 / - Slow Down
Space - Speed up.

 C  - Activate docking computer, if fitted.
 D  - De-activate docking computer if switched on.
 E  - Active ECM, if fitted.
 H  - Hyperspace.
 J  - Warp Jump.
 M  - Fire missile.
 T  - Target a missile.
 U  - Un-target missile.
TAB - Detonate energy bomb, if fitted.
CTRL+H - Galactic Hyperspace, if fitted.
ESC - Launch escape capsule, if fitted.
 P  - Pause game.
 R  - Resume game when paused.

On The Chart Screens
--------------------
D - Select a planet and show distance to it.
F - Find planet by name.
O - Return cursor to current planet.
Cursor Keys - Move cross hairs around.


On The Game Settings Screen
---------------------------
From the Options Screen (F11) you can enter the Game Settings Screen. From here you can change
a number of settings that control how the game looks and plays.  Use the cursor keys to select an option
and the Enter/Return key to change it. The options can be saved as default for future games by pressing Enter
while on the Save Settings option (NB this is not necessary if you want to change the settings just for
the current game).  Game settings are held in the newkind.cfg file which should be in the same directory
as the newkind.exe file.



Release 1.0 - Changes since Beta 3.0
====================================

- Rocks, alloys and boulders no longer stop you from engaging the jump drive.

- Stopped tactics routine from being called while on the intro screens.

- Ramming ships was causing little damage to player's ship.  Now fixed.

- Switching views while in space now switches stars around.

- Moved all Allegro specific graphic functions into alg_gfx.c

- Moved keyboard handling routines into keyboard.c

- Moved file handling routines out of alg_main.c into file.c

- Added screen to allow player to set the game options from within the game.
  This means the player no longer has to hack newkind.cfg directly.

- Tidied graphics routines and removed all references to HDC.

- Fixed movement of crosshairs on chart screens so that they are clipped properly.

- Added explosion, cargo cannisters and alloys to the game over screen.

- Added escape capsule sequence.

- Added option for instant dock vs. auto-pilot docking.

- Fixed bug that allowed an enemy craft to launch multiple escape capsules.

- Fixed asteroids so that they move slower and don't try to evade attack.

- Added rock hermits.

- Tweaked the tactics routine again.

- Added limited joystick support (digital only at the moment).

- Added beep sound when missile target is locked.

- Added low pitch beep sound when missile is unarmed.

- Added Energy Low message when enegry levels reach critical.



Beta 3.0 - Changes since Beta 2.1
=================================

- Fixed bug that allowed lasers to fire too rapidly.

- The console was not being updated (i.e no of missiles) after a saved commander was loaded.  Now fixed.

- Added planet auto-select on chart screens ala NES.

- Asteroid mining implemented.

- Exploding ships now release alloys.

- Fixed bug that caused enemy ships to only fire when at close range.

- Thargoids can now appear.

- Witchspace ambush added.

- Anacondas now release worms and sidewinders when attacked.

- Attacked ships can now release escape capsules.

- Objects other than cargo cannisters can now be scooped.

- Corrected condition indicator code.  Was previously showing yellow even after danger had passed.

- Added ability to pause game (P key pauses, R resumes).

- The Cougar (a cloaked ship) can now appear.

- The second mission is now playable.

- Fixed bug that allowed enemy ships to fire from bizare angles.

- Rewrote tactics routines. New code now based on NES Elite.

- Docking computers now use full auto-pilot rather than just instant dock.

- Space station now launches shuttles and transporters bound for the planet.



Beta 2.1 - Changes since Beta 2.0
=================================

- Added code to set legal status to clean and fuel to maximum after using an escape capsule.

- Fixed enemy missile targeting so that other ships don't blow themselves up.

- Fixed speed of enemy ships which were moving too slowly in previous versions.

- Added game speed control option to newkind.cfg.


Beta 2.0 - Changes since Beta 1.1
=================================

- Fixed bug in enemy tactics which caused ships to run away.

- Set the max speed of player's ship to 0.30LM.

- Added code to display version number on options screen.

- Corrected spelling of "Feudal".

- Enabled docking computers (instant dock for the moment).

- Added hyperspace sound.

- Changed code to display number of credits instead of bounty.

- Set hyperspace count down to 15.

- Added Suns, cabin temperature and fuel scooping.

- Added left and right views when in flight.

- Enabled use of Energy Bomb, activated with TAB key.

- Changed scanner slightly, added code to display objects in different colours.

- Fixed warping of stars in rear view.

- Enabled Escape Capsule, use ESC key to abandon ship.

- Enabled Galactic Hyperspace.

- Added use of 'O' key on chart screens.

- Added ability to find planets by name.

- Added Windows icon created by Marcus Buchanan.

- Added code by Thomas Harte to anti-alias lines and improve hidden surface removal.

- Added option to newkind.cfg to enable anti-alias code.

- Added the first secret mission, hunt the Constrictor!

- Fixed equipment buying so that refund is given on existing lasers.

- Changed colour of fuel limit circle from white to green.

- Put some colour on the Local Chart screen.

- Changed look of ship lollipops on the scanner to look more like the original.

- Changed incoming laser sounds, one for hitting shields and one for hitting the hull.

- Fixed asteroids so that they travel in straight lines.

- Changed creation of ships so that the appear at more random locations.

- Removed the Constrictor and the Cougar from the ship parade intro.



Beta 1.1 - Changes since Beta 1.0
=================================

- Fixed bug in load and save routine which caused the player to aways return to Lave.

- Fixed bug in equipment buying screen.  Buying military lasers for the side or rear view
  would cause mining lasers to be fitted instead.

- Renamed files to be all lowercase.  This should ease ports to other OSes.

- Removed #include of windows only allegro files from alg_main.c

- Added #ifdef around use of GFX_DIRECTX in alg_main.c to allow compilation on non Windows OSes.



Work in progress
================

The following features have not yet been implemented but are being currently worked on.

- C64 Elite style scoring.

- Trumbles (from C64/NES Elite).

- View of docking bay after docking.

- Loading of different consoles not yet fully implemented.

- Enhance joystick support, i.e. use of more buttons and analogue control.


Known Problems
==============

- Bits of hidden surfaces on ships sometimes show through.

- Buying more than 255gs of Gold/Platinum doesn't work.
  It didn't in the original Elite either.  Broken as designed.


Have fun!

Christian Pinder.
<christian@newkind.co.uk>
http://www.newkind.co.uk


docs/ReleaseNotes.txt
   +===================+
   | Elite Adventurous |
   +===================+

< archived release notes >

################    --------                    +----------+
# RELEASE NOTE #    v.beta-1 (0x0000)           | 24/04/16 |
################    --------                    +----------+

Changes:
--------
  git white space, tabs -> 4 spaces
  BugFix: Set clip region before blitting scanner

  Add planet data to dump unvierse (dbg)
  BugFix: Release audio assets (Allegro) on exit

  Isolate common Allegro include files

  Support for screen resolutions
    4:3(800 x 600), 16:10(1280 x 800), 16:9(1280 x 720)
  
  BugFix: Validate cfg file
  BugFix: Centralise display in screen/window
  BugFix: Commander credits not saved properly
  BugFix: Joystick button sensitivity on equip ship
  New format dump universe (dbg)

  Compile time option for Port Audio (ALSA) on linux
    - define USE_ALG_AUDIO, USE_ALSA
    - ALSA loads assets/*.raw, not WAVs
    - Moved sample definitions [assets -> gmlb]

  Added linux build (undef WINDOWS)
    - project files (Makefile, run), mint-64 allegro

  BugFix: Clear obc messages on purchase
  Police obc messages in blue
  Add screen for "Modify Ship" [F3 while docked]
  Audio Scanner
  Homogenise bitmap implementation (scanner.bmp)

Known Bugs/Limitations:
-----------------------
  64-bit visual studio build does not link (allegro build is 32-bit)
  Audio assets (wav) are not mastered (DC offset, unbalanced gain)

################    ---------                   +----------+
# RELEASE NOTE #    v.alpha-5 (0x0000)          | 09/04/16 |
################    ---------                   +----------+

Changes:
--------
  Tidied data on planet screen
  Added install instructions
  Renamed config file
  New install file structure: samples (wavs)
    handled uniformly with bmps, midi, fonts
  Less snes planets
  Window Title (non-directx only)
  Only show obc msg for missile if shot
  Break pattern now reflects ship activity
  New bitmaps (scanner, elitetx3)
  Fix: joystick button sensitivity in menus
  Don't display console items when paused
  isolate allegro calls to gamelib.h|c
  Fix: remove erroneous obc messages

Known Bugs:
-----------
  64-bit build does not link (allegro build is 32-bit)
  Audio assets (wav) are not mastered (DC offset, unbalanced gain)
  Allegro build test on linux, no build procedure/project files for Elite

################    ---------                   +----------+
# RELEASE NOTE #    v.alpha-4 (0x0000)          | 29/03/16 |
################    ---------                   +----------+

Changes:
--------
  Make Ergonomic and Aesthetic Additions:                   [erg_aesth]
    on-board computer (equip ship)
    colour (VGA) scanner (equip ship)
    enhanced planet landscape generation
    variable laser sights, color shots, updated fire rates  [lasers]
    closer emulation of C64 flight controls
    reduce fuel scoop rate
    jump failure message and audio
    improved exception handling (loading config and asset files)
    load all assets individually, not from packed .dat file
    remove scanner.cfg (config is now build time constants)
    set volume (build time constant)
    credits and version number
    updated brk pattern (dock/launch, hyperspace)
    color designed console bars
    horizontal line to separate viewport and console
    extended joystick support (equip, stock, pause)
    extra support for x360 controller (newkind.cfg)
    BF: snes colors (introduced by directX fix)
    BF: double beep when acquiring missile target
  Refactor - style                                          [refactor2]
  Fix failure to run using Direct X (newkind.cfg)           [dx]
  Refactor - reduce scope of program objects                [scope]
  Enable "warnings as errors"                               [warnings]
  Add Visual Studio Community 2015 files                    [vs2015]
  Reorganize file structure, see //readme.txt               [move_files]
  Add Allegro source and 32-bit vs2015 dll build            [allegro]

Known Bugs:
-----------
  No binary or install instructions
  64-bit build does not link (allegro build is 32-bit)
  Audio assets (wav) are not mastered (DC offset, unbalanced gain)
  Joystick auto-fire rate can cause problems with menus
  Allegro build test on linux, no build procedure/project files for Elite


src/gfx.h
/**
 *
 * Elite - The New Kind.
 *
 * The code in this file has not been derived from the original Elite code.
 * Written by C.J.Pinder 1999/2000.
 *  adapted by fuirippu 2016
 **/

#ifndef GFX_H
#define GFX_H


//#define RES_512_512
#define RES_800_600


#pragma region Screen resolution constants

#ifdef RES_512_512

#define GFX_SCALE       (2)
#define GFX_X_OFFSET    (0)
#define GFX_Y_OFFSET    (0)
#define GFX_X_CENTRE    (256)
#define GFX_Y_CENTRE    (192)

#define GFX_VIEW_TX     1
#define GFX_VIEW_TY     1
#define GFX_VIEW_BX     509
#define GFX_VIEW_BY     381

#endif
#ifdef RES_800_600

#define GFX_SCALE       (2)

#define GFX_X_CENTRE    (256)
#define GFX_Y_CENTRE    (192)

#define GFX_VIEW_TX     1
#define GFX_VIEW_TY     1
#define GFX_VIEW_BX     509
#define GFX_VIEW_BY     381

#endif

#ifndef GFX_SCALE

#define GFX_SCALE       (1)
#define GFX_X_OFFSET    (0)
#define GFX_Y_OFFSET    (0)
#define GFX_X_CENTRE    (128)
#define GFX_Y_CENTRE    (96)

#define GFX_VIEW_TX     1
#define GFX_VIEW_TY     1
#define GFX_VIEW_BX     253
#define GFX_VIEW_BY     191

#endif
 
#pragma endregion


/// GFX_COL_xxx defines are indices into the software palette, pColours [colours.c]
/// pColours is set to an array of either dx ARGB values, or GDI palette indices
#pragma region Colour defines
#define GFX_COL_BLACK       0
#define GFX_COL_DARK_RED    1
#define GFX_COL_WHITE       2
#define GFX_COL_GOLD        3
#define GFX_COL_RED         4
#define GFX_COL_CYAN        5

#define GFX_COL_GREY_1      6
#define GFX_COL_GREY_2      7
#define GFX_COL_GREY_3      8
#define GFX_COL_GREY_4      9

#define GFX_COL_BLUE_1      10
#define GFX_COL_BLUE_2      11
#define GFX_COL_BLUE_3      12
#define GFX_COL_BLUE_4      13

#define GFX_COL_RED_3       14
#define GFX_COL_RED_4       15

#define GFX_COL_WHITE_2     16

#define GFX_COL_YELLOW_1    17
#define GFX_COL_YELLOW_2    18
#define GFX_COL_YELLOW_3    19
#define GFX_COL_YELLOW_4    20
#define GFX_COL_YELLOW_5    21

#define GFX_ORANGE_1        22
#define GFX_ORANGE_2        23
#define GFX_ORANGE_3        24

#define GFX_COL_GREEN_1     25
#define GFX_COL_GREEN_2     26
#define GFX_COL_GREEN_3     27

#define GFX_COL_PINK_1      28

#define GFX_COL_SNES_102    29
#define GFX_COL_SNES_134    30
#define GFX_COL_SNES_167    31
#define GFX_COL_SNES_213    32
#define GFX_COL_SNES_255    GFX_COL_WHITE
#define GFX_COL_SNES__83    33
#define GFX_COL_SNES_122    GFX_ORANGE_3
#define GFX_COL_SNES_249    34

#define GFX_COL_AA_0        35
// Eight entries for anti-aliasing GFX_COL_AA_0-7...

#define GFX_COL_FUI         43
#define GFX_COL_BAR_ALERT0  44
#define GFX_COL_BAR_ALERT1  45
#define GFX_COL_BAR_SAFE0   46
#define GFX_COL_BAR_SAFE1   47
#define GFX_COL_BAR_SAFE2   48
#define GFX_COL_BAR_MED0    49
#define GFX_COL_BAR_MED1    50

#define GFX_COL_BAR_MED2    51
#define GFX_COL_CARDINAL    52

#define GFX_COL_BRK_00      53
// Eight entries for break pattern GFX_COL_BRK_00-7...
//      GFX_COL_BRK_07      60

#define GFX_COL_MISSILE     61
#define GFX_COL_VIPER       62
#define GFX_COL_PURPLE_1    63
#define GFX_COL_PURPLE_2    64
#define GFX_COL_SCAN_RED    65
#define GFX_COL_SCAN_YEL    66
#define GFX_COL_MINT        67
#define GFX_COL_NAVAJO      68

#define GFX_COL_PSMITH_00   69
#define GFX_COL_PSMITH_01   70
#define GFX_COL_PSMITH_02   71
#define GFX_COL_PSMITH_03   72
#define GFX_COL_PSMITH_04   73
#define GFX_COL_PSMITH_05   74
#define GFX_COL_PSMITH_06   75
#define GFX_COL_PSMITH_07   76

#define GFX_COL_TRICOL_0    77
// Three entries for tricol GFX_COL_TRICOL_0-2...
//      GFX_COL_TRICOL_2    79

#define GFX_COL_FRAC_0D     80
#define GFX_COL_FRAC_1D     81
#define GFX_COL_FRAC_0L     82
#define GFX_COL_FRAC_1L     83

#define GFX_COL_PINK_SHOCK  84
#define GFX_COL_DUN_TAN     85

#define GFX_COL_NEXT        86
///        ^--- GFX_COL_NEXT must be (last colour + 1)
#pragma endregion

/// The software palette (entries are ARGB for dx, 8b palette index for GDI) [colours.c]
#define NUM_COLOURS         (GFX_COL_NEXT)
extern const int *pColours;
void setColours(int use_directx);

/////////////////////////////////////////////////////////////////////////////
void gfx_fast_plot_pixel(int x, int y, int col);
void gfx_plot_pixel(int x, int y, int col);

void gfx_draw_circle(int cx, int cy, int radius, int colour);
void gfx_draw_filled_circle(int cx, int cy, int radius, int colour);

void gfx_draw_colour_line(int x1, int y1, int x2, int y2, int colour);
void gfx_draw_line(int x1, int y1, int x2, int y2);

void gfx_draw_triangle(int x1, int y1, int x2, int y2, int x3, int y3, int colour);
void gfx_draw_rectangle(int tx, int ty, int bx, int by, int colour);

void gfx_display_colour_text(int x, int y, const char *txt, int colour);
void gfx_display_text(int x, int y, const char *txt);
void gfx_display_pretty_text(int x1, int x2, int y, const char *txt);
void gfx_display_centre_text(int y, const char *str, int psize, int colour);

void gfx_draw_sprite(int sprite_no, int x, int y);
/////////////////////////////////////////////////////////////////////////////

void gfx_draw_scanner(void);

void gfx_clear_display(void);
void gfx_clear_text_area(void);
void gfx_clear_area(int tx, int ty, int bx, int by);

void gfx_set_clip_region(int tx, int ty, int bx, int by);

/////////////////////////////////////////////////////////////////////////////
void gfx_start_render(void);

void gfx_render_polygon(int num_points, int *point_list, int face_colour, int zavg);
void gfx_render_line(int x1, int y1, int x2, int y2, int dist, int colour);

void gfx_finish_render(void);


#endif      // #ifndef GFX_H
